### DISCUSSION TITLE
cfbs add a local git repo?

### DISCUSSION THREAD

**Discussion by t3mcvay** [cfengine/core#5634] [Modules]
I'm starting to learn how to migrate existing CFengine policy to a cfbs managed set.  Our policy repo is a local git repo on disk, and we have a use case/need to NOT make it accessible via something like github.  Is it possible to have a local git repo on disk and then have 'cfbs add /path/to/local/policy' actually be able to recognize that /path/to/local/policy is actually a local repo?

**Comment by olehermanse** [comment]
@t3mcvay Hello, thanks for opening a discussion and good question;

I asssume your policy (on disk) is a fork of the [MPF](https://github.com/cfengine/masterfiles)?

We are working on some features to make this transition easier, but one thing to consider is that we want `cfbs` to produce reproducible builds. In your case, this means that we don't want it to just blindly use what is in `~/some/path/to/masterfiles/`, but we want to it have some checksums / version information in addition to just the path. Currently, without waiting for us to add new features, there's a couple of ways to achieve this;

## Git submodules

A submodule is one mechanism to achieve this. The "normal" way would be to put your dependency (the classic policy set) somewhere on a git server, can be GitHub private repo, on-prem instance of GitLab, or something else. You can then add that URL as a git submodule:

```
$ git submodule add https://server/some/repo/url
```

Since you mentioned not using something like GitHub, instead using local files, it is also possible to use local file paths, but due to security aspects, you will have to explicitly enable submodules via local filesystem paths;

```
$ git config --global protocol.file.allow always
$ git submodule add --force ../masterfiles
```

**Warning:** I wouldn't generally recommend doing this - especially if it's your main work machine where you might git clone some projects and then run a Dockerfile from that project for example. That opens you up to the project referencing some local files via paths and then those being transferred to the docker container running potentially untrusted code. Other similar scenarios exist.

For how to work with git submodules, I'd use the official documentation:

https://git-scm.com/book/en/v2/Git-Tools-Submodules

After you've added the submodule, the last part would be adding the module / editing the `cfbs.json`.
See the section at the end for how your CFEngine Build project (`cfbs.json`) should look like.

## Copying in the files explicitly

The other option would be to copy in the files, with something like rsync;

```
$ cfbs init --non-interactive --masterfiles=no
$ rsync -r --exclude '.git*' --exclude '*.md' --exclude '*.org' --exclude 'tests/' ~/some/path/to/masterfiles/ ./masterfiles/ --delete
$ cfbs --non-interactive add ./masterfiles/
```

(Edit the rsync command with the right path and feel free to exclude more things).

### Updating

With this rsync setup, this is how you'd update from the local path:

```
$ rsync -r --exclude '.git*' --exclude '*.md' --exclude '*.org' --exclude 'tests/' ~/some/path/to/masterfiles/ ./masterfiles/ --delete
$ git add ./masterfiles/
$ git commit -m "Updated ./masterfiles from ~/some/path/to/masterfiles/"
```

For convenience, you can add these commands to an alias, or shell script, if you like.

## CFEngine Build project / Build steps

Whether you're using rsync or git submodules, you can edit the `cfbs.json` to do what's needed. For example:

```json
{
  "name": "Example project",
  "description": "Example description",
  "type": "policy-set",
  "git": true,
  "build": [
    {
      "name": "./masterfiles/",
      "description": "Our old masterfiles folder added as a module, synced using rsync",
      "tags": ["local"],
      "added_by": "cfbs add",
      "steps": ["copy ./ ./"]
    }
  ]
}
```

Or, if your masterfiles repo requires running some script (like `prepare.sh`):

```json
{
  "name": "Example project",
  "description": "Example description",
  "type": "policy-set",
  "git": true,
  "build": [
    {
      "name": "./masterfiles/",
      "description": "Our old masterfiles folder added as a module, synced using rsync",
      "tags": ["local"],
      "added_by": "cfbs add",
      "steps": ["run ./prepare.sh -y", "copy ./ ./"]
    }
  ]
}
```

**Note:** Ensure the rsync command is copying what is needed for your script to run.

## What's next?

We're adding new commands and features to make this nicer for you. Hopefully the workarounds above can work, and we'll release new commands in cfbs soon which will make this easier for you.

When you have a working project, you can use the cfbs tooling and start adding modules from build.cfengine.com. We also recommend you start working towards splitting your classic masterfiles into 2, one part which you maintain (your custom policy files and modifications), and one part which we maintain (the original, versioned, MPF).

**Comment by t3mcvay** [comment]
Hello @olehermanse 

Thanks for the quick reply.  No, your initial assumption is incorrect.   We have a long history with CFengine (way back to an early 2.X version).  We've been carrying that initial custom policy along to now, and do not have any version of the MPF in our CM repo.  That 'old' policy works OK with CFengine 3 (currently working with 3.10). 

But, we think it's time to at least see if incorporating some version of the MPF is feasible at this point.

The current thought is to migrate (1-by-1) our existing policy into the equivalent custom policy (as I understand right now, which would fit under ./services/).  Also -- if something like the submodules example you so kindly provided could be used we could not have a URL as many of our deployed systems are airgapped and we literally just use policy stored in a local repo to administer and setup hosts on a network.  Also I should mention that we have developed (over the years) tools to assist in administering this (such as dynamic generation of things like sudoers, and other custom configurations per host depending upon their function), and that needs to continue as that allows us to write fairly generic policy which will work across a variety of network and security postures.

**Reply by nickanderson** [reply]
Interesting stuff. I wonder if cfbs would help you with managing those various air-gapped policy sets.

How much of that policy can't leave the air-gapped network? Or maybe a better question is much of the policy is specific to individual air gapped networks? 

Would it be easier to use cfbs to build a policy set on your normal network and then sneaker-net the built result over to the air gapped network

**Reply by t3mcvay** [reply]
Hi @nickanderson 

See - that's the $64 question we're trying to answer.  We do fine now with our custom policy set and (most generally) use the same policy across all of our networks as much as possible.  We can do this as we dynamically extract data (IPs, hostnames, etc...) from a dataset which is unique per network.  So the policy itself is pretty generic, but the 'dereferenced variables' may not be, if that makes sense.   Given how we deploy to our airgapped newtworks the answer is: that is also what we're trying to figure out.  I think there may be some appetite for using cfengine vetted modules from github, but not just random things found across the interwebs.

Cheers.

**Reply by nickanderson** [reply]
I think I understand.

Do you have any policies that cant exist off gapped network?

Do you have any dynamic inputs (e.g. loading a policy file based on some variable)?

Do you have have any unique input file names per gapped network (this policy file is only used on gapped network a)?

**Reply by t3mcvay** [reply]
Policy that cannot exist off gapped?  No.  Data that cannot exist off gapped? Yes.

Our entire CFengine policy application is based off of dynamic input - hosts that have defined class membership, which triggers specific bundles and other tools to generate configuration, etc... 

I do not believe we have any uniquely named policy files per network, but for sure unique naming applies to other configuration files.

I am having a hard time wrapping my head around the notion that using something like https://github.com/myproject guarantees a reproducible build whereas file:///local/path/myproject/.git would not.  It seems those two could/should produce the same build (assuming no mods to the repos, of course) - ?

If this would help, this is our current policy deployment flow.  Our policy files + associated data are housed in a local git repo (with possibility of additional data on airgapped networks). After some basic validation/tests (syntax checks, sanity, etc..), that is turned into a RPM via CI/CD.  That RPM is deployed to policy hosts, and trickles to clients as the normal cf-agent cycle runs.

**Reply by olehermanse** [reply]
> I am having a hard time wrapping my head around the notion that using something like https://github.com/myproject guarantees a reproducible build whereas file:///local/path/myproject/.git would not. It seems those two could/should produce the same build (assuming no mods to the repos, of course) - ?

Yes, your thinking is correct. It was more an FYI / what we have to consider when adding support for local file paths. Some clarifications;

- This is only true as long as there is a .git folder (with version information). For https://github.com/username/repo this is pretty much guaranteed. For local files, not so much.
- Our tooling (cfbs and cfbs.json) currently supports: HTTPS URL + version information, and CFEngine Build index module + version information. I.e. we do not have support for the local file path option which you would want.
- As mentioned above, git needs to be configured to allow fetching from local file paths (`git config --global protocol.file.allow always`).

I agree with Nick that it sounds reasonable to work on the project and generate (`cfbs build`) the policy set off the air gapped network, and then transfer the built policy-set to be deployed on the air gapped network.

Apart from that, I can only say that we are working on this and will introduce features to make this easier (especially transitioning from older policy set to cfbs).

**Reply by nickanderson** [reply]
Yeah, it sounds like cfbs could be useful for you even though you don't have per network inputs (i mean uniquely named policy files that need to be specified in inputs of body common control). 

On the data side of things you may want to think about the `augments` key in def.json (you could for example load an augments file that is network specific that lives outside of the policy) as well as host_specific.json (generally you would generate those centrally and distribute them to client) though it sounds like host_specific.json probably would't provide much benefit to you and might just be more work than it's worth.

**Reply by t3mcvay** [reply]
I'm in the middle of slowly figuring out how tot integrate our existing policy in a toy environment  (a policy hub and 1 client).  I need to seriously dig into what a MPF provides / can provide.  I don't think long term we want to give up any of our custom policy unless we find something in the MPF that provides the same functionality/service at less maintenance effort.

In terms of host specific data.  We have adopted the opposite approach.  We centrally maintain a set of XMLs which describes everything about the totality of users, RPMs across all OSes, network, and hosts and then generate the host specific data at the point of use.  Effectively each host ends up with a very customized set of configuration files, right down to things like very unique sudoers files.  It just isn't feasible to maintain the 1000's of individual host files centrally in a sane way (that we've figured out)

**Reply by nickanderson** [reply]
> In terms of host specific data. We have adopted the opposite approach.

Good. Externalizing data like that is generally what we recommend.

> I don't think long term we want to give up any of our custom policy unless we find something in the MPF that provides the same functionality/service at less maintenance effort.

Yeah, generally I wouldn't recommend trying to abandon a system that works well for you. Integrating with the MPF can be beneficial for on-boarding new users that might have some expectations from the docs or conversations with other users.

Mostly I was just curious how you were putting things together and how different the policy sets from each environment were. It sounds to me like your policy sets between environments are perhaps even identical (just consuming different data that changes the behavior).  Where you have actually different policy sets cfbs could be useful to help you keep common parts in sync. So, for you, cfbs probably provides little benefit on that front.  If you are using the MPF then cfbs can make keeping that up to date easier, and it can also be useful to integrate and keep other policies written by others up to date.

**Reply by t3mcvay** [reply]
@nickanderson   -- I think all may become clearer soon. 

To try and expand a bit.  We use a set of XMLs containing host, network, user, and package data per network.  Our policy files contain bundle sets that apply based upon managed class membership, and so are dynamic across the network, but mostly static across networks (if that makes sense).  While we centrally keep the data (e.g. XMLs) we distribute a set of them to each client along with the CF + associated files. These, coupled with externally managed tools which are distributed via a scheduled package update, allows us to provide custom configurations and bundle sets for each host/client.  While many hosts have the same class membership and so end up with ~equivalent configuration sets, there are also man that are unique in terms of functionality (think something like a httpd server + print server in one box).  So all of the stuff we keep in CM is 1 or 2 steps away from the actual configuration files and data that the OS systems/services uses.

I have played with trying to integrate/port some legacy CF to CFBS and found a few things that I'm not sure on.  For example, I wrote a custom promise type for one of the integral tools we use, but find that there is no way to add that to cfbs as no support for python.   I fiddled with the JSON and figured out how to get CFBS to include it in the build and subsequently run a bundle with the custom promise type.  But it seemed harder than it needed to be, and so I am sure I did it wrong.  Similarly, we have quite a few of the 'old definition' module files (e.g. those under ./input/modules) including some custom ones.  I ran out of time, but also could not see a way to natively add those to CFBS without manually munging the JSON.

**Reply by nickanderson** [reply]
Your XMLs sound a bit similar to the Group data in Mission Portal. Beyond the actual format of the data (we use JSON instead of XML) there is a difference in what data is delivered where. We store the Group data in PostgreSQL and generate host specific JSON files and each host looks for it's own host specific data file as part of policy update. While implemented differently they each look to separate data from the policy so that you can affect change without having to modify any policy itself (.cf files).

We have a function, `classfiltercsv()` which might be interesting to you. The point there is to make extracting data that is relevant for the executing context. It's faster than loading a data structure and iterating over it to further filter the relevant data. I don't know that it would improve on what you are doing, but it's at least something to know about.

At any rate, I would not look to try and change any of that in the near term and instead just focus on integrating on top of the MPF.

> but find that there is no way to add that to cfbs as no support for python ...
> could not see a way to natively add those to CFBS without manually munging the JSON.

Needing to manually munge the `cfbs.json` should not be taken as an indication that you did something wrong. In fact this is I think currently required in order to place things in specific places. Adding custom modules is a good example of this. If you `cfbs add ./path/to/my-module.sh` it's going to want to place that inside services. To get it to place it inside the modules directory you do have to modify the `cfbs.json` directly and tweak the path.

**Comment by basvandervlies** [comment]
Just to add we use the augments and  "def.json" setup a lot in conjunction with  [SCL](https://github.com/basvandervlies/cf_surfsara_lib/l).  I have written blog post about our initial setup : https://basvandervlies.blogspot.com/2018/09/cfengine-312-new-features-missingok-and.html#more

Maybe it can help you for your redesign.  Just to add we have completely switch to the cfbs setup and started with cfengine 1.3.X.  So also had a lot of legacy code.

**Comment by t3mcvay** [comment]
@basvandervlies   Thanks for that!!
